apps/designer/app/routes/designer/$id.tsx

```diff
@@ -1,30 +1,69 @@
 import { useLoaderData } from "@remix-run/react";
-import { LoaderFunction } from "@remix-run/node";
+import {
+  ActionFunction,
+  LoaderFunction,
+  unstable_createFileUploadHandler,
+  unstable_parseMultipartFormData,
+} from "@remix-run/node";
 import type { Project } from "@webstudio-is/react-sdk";
 import { Designer, links } from "~/designer";
 import * as db from "~/shared/db";
 import config from "~/config";
 import env from "~/env.server";
+import { type Asset } from "@webstudio-is/react-sdk";
 export { links };
 
 export const loader: LoaderFunction = async ({ params }) => {
   if (params.id === undefined) throw new Error("Project id undefined");
   const project = await db.project.loadById(params.id);
+  const assets = await db.assets.loadByProject(params.id);
   if (project === null) {
     return { errors: `Project "${params.id}" not found` };
   }
-  return { config, project, env };
+  return { config, assets, project, env };
 };
 
+const directory = "./public/uploads";
+
 type Data = {
   config: typeof config;
   project: Project;
+  assets: Asset[];
 };
 
 type Error = {
   errors: "string";
 };
 
+type ImageUpload = {
+  type: string;
+  name: string;
+};
+
+export const action: ActionFunction = async ({ request, params }) => {
+  if (params.id === undefined) throw new Error("Project id undefined");
+  const formData = await unstable_parseMultipartFormData(
+    request,
+    unstable_createFileUploadHandler({
+      maxPartSize: 10_000_000,
+      directory,
+      file: ({ filename }) => filename,
+    })
+  );
+  const imageInfo = formData.get("image") as ImageUpload;
```

Comment:
I made this mistake too, instead of telling typescript it is ImageUpload as a truth, we should be actually asserting it is the case with code, because this line is already wrong, because it can be null, but also it can be something else, better check would be:

```ts
if (imageInfo instanceof ImageUpload === false) return {errors: 'what happened'}
```

As we speak there must be a way to handle an error in the UI for this

Reply:
I got a reccomendation lately for this https://github.com/esamattis/react-zorm

Reply:
I pushed something, let me know how to improve it